Library/Homebrew/cmd/--env.rb

```diff
@@ -1,20 +1,128 @@
 #:  * `--env`:
 #:    Show a summary of the Homebrew build environment.
 
+
 require "extend/ENV"
 require "build_environment"
 
+
 module Homebrew
+  
+  MAX_ITERATIONS = 10
+  VALID_SHELLS = %w(
+    sh csh tcsh bash ksh mksh oksh
+    pdksh zsh fish
+  )
+  DEFAULT_SHELL = "sh"
+
+
+  def canonicalize_shell(path)
+    # basename and strip leading '-' in login shell
+    name = File.basename path
+
+    if VALID_SHELLS.include? name
+      return name
+    end
+    if name.start_with? "-"
+      stripped_name = name.sub(/^[-]/, "")
+      if VALID_SHELLS.include? stripped_name
+        return stripped_name
+      end
+    end
+    return nil
+  end
+
+
+  def get_explicit_shell
+    # parse shell from command line arguments
+    split_arg_index = ARGV.index "--shell"
+    if split_arg_index
+      if split_arg_index + 1 < ARGV.length
+        return ARGV[split_arg_index + 1]
+      else
+        STDERR.puts "--shell option requires argument"
+        exit
+      end
+    end
+
+    long_arg_index = ARGV.index {|s| s.start_with? "--shell="}
+    if long_arg_index
+      shell_name = ARGV[long_arg_index].sub(/^[-][-]shell=/, "")
+      if shell_name == ""
+        return ""
+      end
+
+      canonical = canonicalize_shell(shell_name)
+      if canonical
+        return canonical
+      else
+        STDERR.puts "argument #{shell_name} to --shell option is not a valid shell"
+        exit
+      end
+    end
+    
+    return nil
+  end
+
+
+  def get_ancestor_shell
+    # get the command name for the
+    # nearest parent, grandparent, etc. process
+    # that looks like a shell
+    candidate_pid = Process.ppid.to_s
+
+    for _ in 1..MAX_ITERATIONS
+      candidate_command = `ps -p #{candidate_pid} -o command=`.chomp
+      candidate_shell = canonicalize_shell(candidate_command)
+      if candidate_shell
+        return candidate_shell
+      end
+
+      candidate_pid = `ps -p #{candidate_pid} -o ppid=`.chomp
+      if candidate_pid == ""
+        break
+      end
+    end
+    return nil
+  end
+
+
   def __env
     ENV.activate_extensions!
     ENV.deps = ARGV.formulae if superenv?
     ENV.setup_build_environment
     ENV.universal_binary if ARGV.build_universal?
 
-    if $stdout.tty?
+    explicit_shell = get_explicit_shell()
+
+    if explicit_shell
+      shell = explicit_shell
+    elsif $stdout.tty?
+      shell = nil
+    else
+      ancestor_shell = get_ancestor_shell()
+      if ancestor_shell
+        shell = ancestor_shell
+      else
+        shell = DEFAULT_SHELL
+      end
```

Comment:
Shorter: `shell = get_ancestor_shell || DEFAULT_SHELL`. Also you might want to rename `get_ancestor_shell` to `ancestor_shell`.